app/src/main/java/com/eveningoutpost/dexdrip/receivers/aidex/AidexReceiver.java

```diff
@@ -0,0 +1,311 @@
+            return;
+        }
+
+        if (bgValueMgDl <= 0) {
+            UserError.Log.w(TAG, "Aidex Broadcast received BG=0, ignoring it.");
+            return;
+        }
+
+        checkIfCorrectSensorIsRunning(sensorId, timeStamp);
+
+        UserError.Log.i(TAG, "Aidex Broadcast NewBGEstimate received: bg=" + bgValueMgDl + ", time=" + JoH.dateTimeText(timeStamp));
+        BgReading.bgReadingInsertFromInt(bgValueMgDl, timeStamp, segmentation_timeslice, false);
+
+    }
+
+
+    private void checkIfCorrectSensorIsRunning(String sensorId, long timeStamp) {
+        Sensor currentSensor = Sensor.currentSensor();
+        if(currentSensor!=null) {
+            if (!currentSensor.uuid.equals(sensorId)) {
+                Sensor.stopSensor();
+                Sensor.create(timeStamp - 10000, sensorId);
+            }
+        } else {
+            Sensor.create(timeStamp - 10000, sensorId);
+        }
+    }
+
+
+    private void processCalibration(Bundle bundle) {
+        if (Pref.getBooleanDefaultFalse("accept_broadcast_calibrations")) {
+
+            final long calibration_timestamp = bundle.getLong(AIDEX_TIMESTAMP, -1);
+            double bgValue = bundle.getDouble(AIDEX_BG_VALUE, -1);
+            final String units = bundle.getString(AIDEX_BG_TYPE, "mg/dl");
+
+            final long timeoffset = JoH.tsl() - calibration_timestamp;
+
+            if (bgValue > 0) {
+
+                if (timeoffset < 0) {
+                    Home.toaststaticnext(gs(R.string.got_calibration_in_the_future__cannot_process));
+                    return;
+                }
+
+                final String local_units = Pref.getString("units", "mgdl");
+                if (AidexBroadcastIntents.UNIT_MG_DL.equals(units) && (!local_units.equals("mgdl"))) {
+                    bgValue = bgValue * Constants.MGDL_TO_MMOLL;
+                    Log.d(TAG, "Converting from mgdl to mmol: " + JoH.qs(bgValue, 2));
+                } else if (AidexBroadcastIntents.UNIT_MMOL_L.equals(units) && (!local_units.equals("mmol"))) {
+                    bgValue = bgValue * Constants.MMOLL_TO_MGDL;
+                    Log.d(TAG, "Converting from mmol to mgdl: " + JoH.qs(bgValue, 2));
+                }
+
+                UserError.Log.ueh(TAG, "Processing broadcasted calibration: " + JoH.qs(bgValue, 2) + " offset ms: " + JoH.qs(timeoffset, 0));
+                final Intent calintent = new Intent(xdrip.getAppContext(), AddCalibration.class);
+                calintent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
+                calintent.putExtra("timestamp", JoH.tsl());
+                calintent.putExtra("bg_string", JoH.qs(bgValue));
+                calintent.putExtra("bg_age", Long.toString(timeoffset / 1000));
+                calintent.putExtra("allow_undo", "true");
+                calintent.putExtra("note_only", "false");
+                calintent.putExtra("cal_source", AIDEX_RECEIVER);
+                Home.startIntentThreadWithDelayedRefresh(calintent);
```

Comment:
You are right. My knowledge of Xdrip is limited, and I found this code, and I thought this is the way to do it. 
But you are right, we get already processed data.

Please let me know how to fix it and I will do it.

Reply:
You can do like ` BloodTest.create(timestamp, mgdl, "Aidex", null)` - Timestamp is milliseconds since epoch, for now use `JoH.tsl()` and mgdl is glucose in mgdl.

Reply:
fixed.